packages/types/src/type/ExtrinsicEra.ts

```diff
@@ -2,33 +2,189 @@
 // This software may be modified and distributed under the terms
 // of the Apache-2.0 license. See the LICENSE file for details.
 
+import Enum from '../codec/Enum';
+import { assert, hexToU8a, isHex, isU8a, isObject, u8aToBn } from '@polkadot/util';
+import Tuple from '../codec/Tuple';
+import U8 from '../primitive/U8';
 import { AnyU8a } from '../types';
-
-import { u8aToU8a } from '@polkadot/util';
-
 import U8a from '../codec/U8a';
 
 /**
  * @name ExtrinsicEra
  * @description
  * The era for an extrinsic, indicating either a mortal or immortal extrinsic
  */
-export default class ExtrinsicEra extends U8a {
+export default class ExtrinsicEra extends Enum {
+  constructor (value?: any) {
+    super({ ImmortalEra, MortalEra }, ExtrinsicEra.decodeExtrinsicEra(value));
+  }
+
+  private static decodeExtrinsicEra (value: MortalMethod | Uint8Array | string): Uint8Array | Object | undefined {
+    if (!value) {
+      return ExtrinsicEra.decodeExtrinsicEra(new Uint8Array([0]));
+    }
+    if (isHex(value)) {
+      return ExtrinsicEra.decodeExtrinsicEra(hexToU8a(value));
+    } else if (isU8a(value)) {
+      if (value[0] === 0) {
+        return new Uint8Array([0, 0]);
+      } else {
+        return new Uint8Array([1, value[0], value[1]]);
+      }
+    } else if (isObject(value)) {
+      return { MortalEra: value };
+    }
+    throw new Error('Invalid data passed to Era');
+  }
+
+  /**
+   * @description Returns the item as a [[ImmortalEra]]
+   */
+  get asImmortalEra (): ImmortalEra {
+    assert(this.isImmortalEra, `Cannot convert '${this.type}' via asImmortalEra`);
+    return this.value as ImmortalEra;
+  }
+
+  /**
+   * @description Returns the item as a [[MortalEra]]
+   */
+  get asMortalEra (): MortalEra {
+    assert(this.isMortalEra, `Cannot convert '${this.type}' via asMortalEra`);
+    return this.value as MortalEra;
+  }
+
+  /**
+   * @description `true` if Immortal
+   */
+  get isImmortalEra (): boolean {
+    return this.index === 0;
+  }
+
+  /**
+   * @description `true` if Mortal
+   */
+  get isMortalEra (): boolean {
+    return this.index > 0;
+  }
+
+  /**
+   * @description Encodes the value as a Uint8Array as per the parity-codec specifications
+   * @param isBare true when the value has none of the type-specific prefixes (internal)
+   */
+  toU8a (isBare?: boolean): Uint8Array {
+    return this.value.toU8a();
+  }
+}
+
+const VALID_IMMORTAL = new U8a([0]);
+/**
```

Comment:
(nit) Blank line between the const and the comment. Even better would be to move the `const VALID_IMMORTAL` to the top of the file, just before the first class definition so all global constants are visible there.